text/0000-build-systems.md

```diff
@@ -0,0 +1,446 @@
+| Build configuration   | Cargo settings per crate in graph                                                                                                                        | Profiles                            |
+| Build lowering        | A build plan: a series of steps that must be run in sequence, including rustc and binary invocations | Build scripts, plugins              |
+| Build execution       | Compiled artifacts                                                                                                                                       | Caching                             |
+
+The first stage, dependency resolution, is the most complex; it's where our
+model of semver comes into play, as well as a huge list of related
+concerns. Interestingly, it's also the area where external build systems tend to
+have the *least* to say, because it's not really about *building* so much as
+*figuring out what even goes into a build*.
+
+Dependency resolution produces a lockfile, which records what crates are
+included in the dependency graph, coming from what sources and at what versions,
+as well as interdependencies. It operates independently of the requested Cargo
+workflow.
+
+The next stage is build *configuration*, which conceptually is where things like
+profiles come into play: of the crates we're going to build, we need to decide,
+at a high level, "how" we're going to build them. This configuration is at the
+"Cargo level of abstraction", i.e. in terms of things like profiles rather than
+low-level rustc flags. There's strong desire to make this system more
+expressive, for example by allowing you to always optimize certain dependencies
+even when otherwise in the debug profile.
+
+After configuration, we know at the Cargo level exactly what we want to build,
+but we need to *lower* the level of abstraction into concrete, individual
+steps. This is where, for example, profile information is transformed into
+specific rustc flags. Lowering is done independently for each crate, and results
+in a sequence of process invocations, interleaving calls to `rustc` with
+e.g. running the binary for a build script. You can think of these sequences as
+expanding what was previously a "compile this crate with this configuration"
+node in the dependency graph into a finer-grained set of nodes for running rustc
+etc.
+
+Finally, there's the actual build *execution*, which is conceptually
+straightforward: we analyze the dependency graph and existing, cached artifacts,
+and then actually perform any un-cached build steps (in parallel when
+possible). Of course, this is the bread-and-butter of many external build
+systems, so we want to make it easy for them to tweak or entirely control this
+part of the process.
+
+The first two steps -- dependency resolution and build configuration -- need to
+operate on an entire dependency graph at once. Build lowering, by contrast, can
+be performed for any crate in isolation.
```

Comment:
This could be turned into a column of the table for clarity. "Scope" -> "Entire dependency graph"/"Individual crate"